src/metabase/driver/sql/query_processor/boolean_to_comparison.clj

```diff
@@ -0,0 +1,104 @@
+(ns metabase.driver.sql.query-processor.boolean-to-comparison
+  "In Oracle and some other databases, boolean literals cannot appear in the top-level of WHERE clauses or expressions
+  like AND, OR, NOT, and CASE. These instead require a comparison operator, so boolean constants like 0 and 1 must be
+  replaced with equivalent expressions like 1 = 1 or 0 = 1.
+
+  Drivers can call boolean->comparison to convert boolean literals and refs into comparison expressions. See the
+  sqlserver or oracle drivers for examples."
+  (:require
+   [metabase.driver.sql.query-processor :as sql.qp]
+   [metabase.legacy-mbql.util :as mbql.u]
+   [metabase.lib.metadata :as lib.metadata]
+   [metabase.query-processor.store :as qp.store]))
+
+;; Oracle and SQLServer (and maybe others) use 0 and 1 for boolean constants, but, for example, none of the following
+;; queries are valid in such databases:
+;;
+;; - SELECT 1 WHERE 1
+;; - SELECT 1 WHERE 1 AND 1
+;; - SELECT CASE WHEN 1 THEN 1 ELSE 0 END
+;;
+;; But these are:
+;;
+;; - SELECT 1 WHERE (1 = 1)
+;; - SELECT 1 WHERE (1 = 1) AND (1 = 1)
+;; - SELECT CASE WHEN (1 = 1) THEN 1 ELSE 0 END
+;;
+;; https://learn.microsoft.com/en-us/sql/t-sql/data-types/constants-transact-sql#boolean-constants
+;; https://learn.microsoft.com/en-us/sql/t-sql/language-elements/comparison-operators-transact-sql#boolean-data-type
+;; https://learn.microsoft.com/en-us/sql/t-sql/language-elements/and-transact-sql
+;;
+;; https://docs.oracle.com/en/database/oracle/oracle-database/21/sqlrf/About-SQL-Conditions.html#GUID-E9EC8434-CD48-4C01-B01B-85E5359D8DD7
+;; https://docs.oracle.com/en/database/oracle/oracle-database/21/sqlrf/Data-Types.html#GUID-285FFCA8-390D-4FA9-9A51-47B84EF5F83A
+;; https://docs.oracle.com/en/database/oracle/oracle-database/21/sqlrf/Logical-Conditions.html
+
+(def ^:private default-boolean-types #{:type/Boolean})
+
+(defn- some-isa? [child parents]
+  (some #(isa? child %) parents))
+
+(defn- boolean-typed?
+  ([m]
+   (boolean-typed? m default-boolean-types))
+  ([m boolean-types]
+   (and (map? m)
+        (some-isa? ((some-fn :base-type :effective-type)
+                    ;; :value clauses have snake keys like :base_type, but field metadata is a snake-hating-map and
+                    ;; will throw if you try to access snake keys, so normalize them first.
+                    (update-keys m mbql.u/normalize-token))
+                   boolean-types))))
+
+(defn- boolean-typed-clause? [[_tag _x options]]
+  (boolean-typed? options))
+
+(defn- boolean-field-clause? [clause boolean-types]
+  (and (mbql.u/is-clause? :field clause)
+       (let [[_ id-or-name options] clause
+             has-types? (some-fn :base-type :base_type :effective-type :effective_type)]
```

Comment:
Why is this not top-level and why is it plural?

Reply:
No reason it couldn't be top-level, probably because I only needed it locally in this function.

As for the name... good question. Personally, I would expect a predicate named `has-type?`  to check a *specific* type, like `(has-type? foo :type/Boolean)` or something. To my mind, the plural `has-types?` conveys that I'm not interested in the specific type, only that *some* type is present (both in the sense that I don't care if it's `:base-type` or `:effective-type`, and in the sense that I don't care about the actual concrete type), but that might just be me. Maybe a better name would be `has-type-option?` or `has-some-type?`.